Rewrite/update compiler source code chapter (#765)
rust-lang/rustc-dev-guide: src/compiler-src.md

on the latter.

Rustc consists of a number of crates, including `rustc_ast`,
`rustc`, `rustc_target`, `rustc_codegen`, `rustc_driver`, and
many more. The source for each crate can be found in a directory
like `src/libXXX`, where `XXX` is the crate name.

(N.B. The names and divisions of these crates are not set in
stone and may change over time. For the time being, we tend towards a
finer-grained division to help with compilation time, though as incremental
compilation improves, that may change.)

The dependency structure of these crates is roughly a diamond:

```text
                  rustc_driver
                /      |       \
              /        |         \
            /          |           \
          /            v             \
rustc_codegen  rustc_borrowck   ...  rustc_metadata
          \            |            /
            \          |          /
              \        |        /
                \      v      /
                  rustc_middle
                       |
                       v
                   rustc_ast
                    /    \
                  /       \
           rustc_span  rustc_builtin_macros

> ```

Reviewer: This is a really neat diagram, could we keep it (with crate names updated)?

Reviewer 2: In the last month or so I believe that there was discussion about mdbook support for diagrams.  Has this been implemented? We could work in a number of diagrams across the guide.

Author: I think it's more likely for us to use something like mermaid. I believe the chalk book does that.

The hard part is that the crate graph is rather complicated and no longer looks like the original graph.

So I would like to try to merge this and add useful diagrams later.

Reviewer: > The hard part is that the crate graph is rather complicated and no longer looks like the original graph.

I see, that's unfortunate.

reporting mechanism. They are also an internal part
of the compiler and not intended to be stable (though they do wind up
getting used by some crates in the wild; a practice we hope to
gradually phase out).

## The main stages of compilation

The Rust compiler is in a bit of transition right now. It used to be a
purely "pass-based" compiler, where we ran a number of passes over the
entire program, and each did a particular check of transformation. We
are gradually replacing this pass-based code with an alternative setup
based on on-demand **queries**. In the query-model, we work backwards,
executing a *query* that expresses our ultimate goal (e.g. "compile
this crate"). This query in turn may make other queries (e.g. "get me
a list of all modules in the crate"). Those queries make other queries
that ultimately bottom out in the base operations, like parsing the
input, running the type-checker, and so forth. This on-demand model
permits us to do exciting things like only do the minimal amount of
work needed to type-check a single function. It also helps with
incremental compilation. (For details on defining queries, check out

> the [query model].)

Reviewer: I think we should keep this paragraph.

Reviewer: Also, if we need examples of _exciting_ things with queries, I've been doing lots of fancy things for https://github.com/rust-lang/rust/pull/73566 (e.g. [not running typeck at all](https://github.com/rust-lang/rust/pull/73566/files#diff-0de644786a2f1fd88c7d7f44bc3fa4bbR387)).

Author: @jyn514 I feel like this paragraph is redundant to a few other places in the guide (e.g. the overview chapter or the queries chapter), whereas this chapter is more about the structure of the code itself.

Regarding the example, I think it could work well in one of the above mentioned chapters.

Reviewer: Ah good point, I forgot that there were two whole chapters just for queries. In that case removing this seems fine.

    - An important step in processing the HIR is to perform type
      checking. This process assigns types to every HIR expression,
      for example, and also is responsible for resolving some
      "type-dependent" paths, such as field accesses (`x.f` – we
      can't know what field `f` is being accessed until we know the
      type of `x`) and associated type references (`T::Item` – we
      can't know what type `Item` is until we know what `T` is).
    - Type checking creates "side-tables" (`TypeckTables`) that include
      the types of expressions, the way to resolve methods, and so forth.
    - After type-checking, we can do other analyses, such as privacy checking.
4. **Lowering to MIR and post-processing**
    - Once type-checking is done, we can lower the HIR into MIR ("middle IR"),
      which is a **very** desugared version of Rust, well suited to borrowck
      but also to certain high-level optimizations.
5. **Translation to LLVM and LLVM optimizations**
    - From MIR, we can produce LLVM IR.
    - LLVM then runs its various optimizations, which produces a number of
      `.o` files (one for each "codegen unit").
6. **Linking**

>     - Finally, those `.o` files are linked together.

Reviewer: I like this step-by-step summary. Is there something like it elsewhere in the guide? If so we should link there, if not I think we should keep this.

Author: Yes, there is this: https://rustc-dev-guide.rust-lang.org/overview.html

I will add a link.

and [`src/tools/rustdoc-themes`].

You can read more about rustdoc in [this chapter][rustdocch].

[`librustdoc`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/index.html
[`rustdoc::main`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/fn.main.html
[`src/tools/rustdoc`]:  https://github.com/rust-lang/rust/tree/master/src/tools/rustdoc
[`src/tools/rustdoc-js`]: https://github.com/rust-lang/rust/tree/master/src/tools/rustdoc-js
[`src/tools/rustdoc-themes`]: https://github.com/rust-lang/rust/tree/master/src/tools/rustdoc-themes

> [rustdocch]: ./rustdoc-internals.md

Reviewer: I think it would be better to link to the rustdoc overview than the specifics of how things work.
```suggestion
[rustdocch]: ./rustdoc.md
```

[main]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/fn.main.html
[`rustc_driver`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/index.html
[`rustc_interface`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_interface/index.html
[`rustc_middle`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/index.html
[`rustc_data_structures`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_data_structures/index.html
[`rustc_span`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/index.html
[`Span`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/struct.Span.html
[`rustc_errors`]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/index.html

You can see the exact dependencies by reading the `Cargo.toml` for the various
crates, just like a normal Rust crate.

One final thing: [`src/llvm-project`] is a submodule for our fork of LLVM.

Most of this book is about the compiler, so we won't have any further
explanation of these crates here.

[`src/llvm-project`]: https://github.com/rust-lang/rust/tree/master/src

### Big picture

The dependency structure is influenced strongly by two main factors:

> 1. Organization. The compiler is a _huge_ codebase; it would be an impossibly

Reviewer: Out of curiosity I ran tokei on it - 1.7 _million_ lines of code.

Author: Did you exclude LLVM? That sounds a bit higher than I remember (~500K), but I might be wrong.

Reviewer: That was only counting lines of rust.

<details>

```
-------------------------------------------------------------------------------
 Language            Files        Lines         Code     Comments       Blanks
-------------------------------------------------------------------------------
 Alex                    4           85           67            0           18
 Assembly                3         1375          756          402          217
 GNU Style Assembly   6740      1516071       721657       473426       320988
 Autoconf              354        38316        22408        10153         5755
 Automake               12         6065         5149          162          754
 BASH                   10         1791         1387          226          178
 Batch                  18         1598         1342           50          206
 C                    6315       934950       529392       304222       101336
 C Header             7670      1448926       897283       335737       215906
 CMake                1196        59281        47496         5105         6680
 C#                      8          766          571          106           89
 C++                 20444      5295316      3766548       869578       659190
 C++ Header             29        11544         9675          809         1060
 CSS                    43         7378         6041          295         1042
 D                       1           18           16            0            2
 Dockerfile             89         2963         2174          308          481
 .NET Resource           2          269          132          113           24
 Emacs Lisp              9         1296          910          241          145
 FORTRAN Modern          8          398            0          371           27
 GDB Script              1           41           13           15           13
 Go                   1912       482175       362283        71149        48743
 Handlebars              4          138          116            7           15
 HEX                     1           15           15            0            0
 HTML                  186       462384       444556         4440        13388
 INI                     3           30           25            0            5
 JavaScript             55         8000         6409          886          705
 JSON                  144        10028        10028            0            0
 LLVM                23437      3668320      1477481      1865951       324888
 LD Script               1            4            4            0            0
 Lua                     1            7            5            0            2
 Makefile              804         6620         4481          697         1442
 Markdown             1936       171532       171532            0            0
 Module-Definition     118        25860        23250           72         2538
 MSBuild                 8          597          485           88           24
 OCaml                  89        11140         5997         3210         1933
 Objective-C          1721       111733        62368        30723        18642
 Objective-C++         496        37842        24784         6426         6632
 Pascal                 10          316           54          228           34
 Perl                   22         9025         5851         1612         1562
 Protocol Buffers        2          173          118           31           24
 Python               1673       237900       172805        24597        40498
 R                       1           15            7            5            3
 RPM Specfile            1          469          390            0           79
 ReStructuredText     1220       219333       219333            0            0
 Ruby                    1          125           89           23           13
 Rust                19032      1672465      1202002       276415       194048
 Scala                   6           54           54            0            0
 Shell                 246        65387        48964         9122         7301
 SVG                    53         5019         4661          358            0
 Swift                   2           23           17            0            6
 SWIG                   69        10949         7683          512         2754
 TeX                     6         3374         2992           22          360
 Plain Text           1506       517261       517261            0            0
 TOML                  797         8688         7335          179         1174
 TypeScript             25         3784         2906          345          533
 Vim Script             12          593          484           41           68
 Visual Studio Sol|      2           47           47            0            0
 XSL                     3          278          226           14           38
 XML                    51       135907       134752           18         1137
 YAML                  747        73902        59039        12268         2595
-------------------------------------------------------------------------------
 Total               99359     17289959     10993906      4310758      1985295
-------------------------------------------------------------------------------
```

</details>

   large crate. In part, the dependency structure reflects the code structure
   of the compiler.
2. Compile time. By breaking the compiler into multiple crates, we can take
   better advantage of incremental/parallel compilation using cargo. In
   particular, we try to have as few dependencies between crates as possible so
   that we dont' have to rebuild as many crates if you change one.

At the very bottom of the dependency tree are a handful of crates that are used
by the whole compiler (e.g. [`rustc_span`]). The very early parts of the
compilation process (e.g. parsing and the AST) depend on only these.

Pretty soon after the AST is constructed, the compiler's [query system][query]
gets set up.  The query system is set up in a clever way using function
pointers. This allows us to break dependencies between crates, allowing more
parallel compilation.

However, since the query system is defined in [`rustc_middle`], nearly all
subsequent parts of the compiler depend on this crate. It is a really large
crate, leading to long compile times. Some efforts have been made to move stuff

> out of it with limited success. Another unfortunate sideffect is that sometimes

Reviewer: ```suggestion
out of it with limited success. Another unfortunate side effect is that sometimes
```

crates, just like a normal Rust crate.

One final thing: [`src/llvm-project`] is a submodule for our fork of LLVM
During bootstrapping, LLVM is built and the [`src/librustc_llvm`] and
[`src/rustllvm`] crates contain rust wrappers around LLVM (which is written in
C++), so that the compiler can interface with it.

Most of this book is about the compiler, so we won't have any further
explanation of these crates here.

[`src/llvm-project`]: https://github.com/rust-lang/rust/tree/master/src
[`src/librustc_llvm`]: https://github.com/rust-lang/rust/tree/master/src
[`src/rustllvm`]: https://github.com/rust-lang/rust/tree/master/src

### Big picture

The dependency structure is influenced strongly by two main factors:

1. Organization. The compiler is a _huge_ codebase; it would be an impossibly
   large crate. In part, the dependency structure reflects the code structure
   of the compiler.
2. Compile time. By breaking the compiler into multiple crates, we can take
   better advantage of incremental/parallel compilation using cargo. In
   particular, we try to have as few dependencies between crates as possible so

>    that we dont' have to rebuild as many crates if you change one.

Reviewer: ```suggestion
   that we don't have to rebuild as many crates if you change one.
```
